osbenchmark/tuning/optimal_finder.py

```diff
@@ -0,0 +1,169 @@
+# Licensed to Elasticsearch B.V. under one or more contributor
+# license agreements. See the NOTICE file distributed with
+# this work for additional information regarding copyright
+# ownership. Elasticsearch B.V. licenses this file to you under
+# the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#	http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import os
+import sys
+import csv
+import uuid
+import logging
+import tempfile
+import subprocess
+from datetime import datetime
+from timeit import default_timer as timer
+from osbenchmark.tuning.schedule import BatchSizeSchedule, BulkSizeSchedule, ClientSchedule, ScheduleRunner
+from osbenchmark.tuning.result import Result
+
+
+def get_benchmark_params(args, batch_size, bulk_size, number_of_client, temp_output_file):
+    params = {}
+    params["--target-hosts"] = args.target_hosts
+    if args.client_options:
+        params["--client-options"] = args.client_options
+    params["--kill-running-processes"] = None
+    # we only test remote cluster
+    params["--pipeline"] = "benchmark-only"
+    params["--telemetry"] = "node-stats"
+    params["--telemetry-params"] = ("node-stats-include-indices:true,"
+                                    "node-stats-sample-interval:10,"
+                                    "node-stats-include-mem:true,"
+                                    "node-stats-include-process:true")
+    params["--workload-path"] = args.workload_path
+    params["--workload-params"] = get_workload_params(batch_size, bulk_size, number_of_client)
+    # generate output
+    params["--results-format"] = "csv"
+    params["--results-file"] = temp_output_file
+    return params
+
+
+def get_workload_params(batch_size, bulk_size, number_of_client):
+    params = [f"bulk_size:{bulk_size}",
+              f"batch_size:{batch_size}",
+              f"bulk_indexing_clients:{number_of_client}",
+              f"index_name:{generate_random_index_name()}"]
+
+    return ",".join(params)
+
+
+def run_benchmark(params):
+    commands = ["opensearch-benchmark", "execute-test"]
+    for k, v in params.items():
+        commands.append(k)
+        if v:
+            commands.append(v)
+
+    proc = None
+    try:
+        proc = subprocess.Popen(
+            commands,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE)
+
+        _, stderr = proc.communicate()
+        return proc.returncode == 0, stderr.decode('ascii')
+    except KeyboardInterrupt as e:
+        proc.terminate()
+        print("Process is terminated!")
+        raise e
```

Comment:
I understand the convenience of this but at the same time I'm not sure if this is best practice. This makes it feel like more of a wrapper utilizing OSB rather than a built-in feature for OSB.

Reply:
It's similar to `compare` command which also utilizes OSB and can potentially reside outside OSB, but it's also strongly coupled with OSB as it always has to work together with OSB. I feel it would just bring unnecessary churns to user if the command is in its own package that users have to install both to make it work.

Reply:
`compare` subcommand is a light-weight operation that pulls from the test execution store (whether that is local or from an external metric datastore) and does some diffs between them. Essentially, this is still a a recommendation tool that wraps OSB.